### DISCUSSION TITLE
`model(x, training=False)` and `model(x, training=True)` outputs differ

### DISCUSSION THREAD

**Discussion by santi-mir** [keras-team/keras#20204] [Q&A]
I've just [read this issue in tf-keras](https://github.com/keras-team/tf-keras/issues/798), and ended up testing a simple convnet with mnist to understand the problem.

I suspect the reason may be due to BatchNormalization, but don't know enough to understand why.

Here is a simple image comparing the _mean absolute error_ for `model(X, training=True)` and `model.predict(X)` which disappears is `training=False`:

![Screenshot 2024-09-03 203535](https://github.com/user-attachments/assets/08714d59-437a-4808-922b-bca39afd75c2)

But why would that make the outputs different, between `model(X, training=False)` (or `model.predict`), and `model(X, training=True)`

I don't have a colab link, but the snippet I used is:

```py
for step, (x_batch_test, y_batch_test) in enumerate(ds_test):

    p = model(x_batch_test, training=True)
    r = np.average(mae(p, y_batch_test))
    print("model ",r)

    p = model.predict(x_batch_test)
    r = np.average(mae(p, y_batch_test))
    print("model predict ", r)
```
Is there a simple explanation? I should read the code but maybe someone knows already.

Anyone that could help, please?

**Comment by abhaskumarsinha** [comment]
Hello @ghsanti 

Thank you for pointing the issue out. Can you please set `keras.layers.BatchNormalization` argument `epsilon= 1e-9` manually which is by default set to `0.001` and re-check the issue again?

I suspect division by 0 or near-0 numbers are the culprit here.

Best Regards,
Abhas

**Reply by santi-mir** [reply]
I think I found the answer in the docs, but let me know if you think otherwise:

> (...) e.g., model(x), or model(x, training=False) if you have layers such as BatchNormalization that behave differently during inference.

[Source in the docs.](https://keras.io/api/models/model_training_apis/#predict-method)

I'm guessing that BatchNormalization during training computes the _mean_ and _std_ from a batch and applies it, whereas in inference mode it just uses those parameters.
